proposals/text/5629-file-similarity-retriever.md

```diff
@@ -0,0 +1,186 @@
+- primary_retriever: First retriever.
+- secondary_retriever (optional): Second retriever (if applicable).
+- keep_original_score (optional): Set this to store the original relevance score of the returned document in the document's meta field (the documents' scores get replaced in the FileSimilarityRetriever output with the reciprocal rank fusion score).
+- top_k: How many documents to return.
+- max_query_len: How many characters can be in a query document. The documents exceeding this limit will be cut off. This was added because in some cases BM25Retriever threw an error if the query exceeded the `max_clause_count` search setting (https://www.elastic.co/guide/en/elasticsearch/reference/7.17/search-settings.html)
+- max_num_queries (optional): The maximum number of queries that should be run for a single file. If the number of query documents exceeds this limit, the query documents will be split into n parts so that n < max_num_queries and every nth document will be kept.
+- use_existing_embeddings: Whether to re-use the existing embeddings from the index. To optimize speed for the file similarity retrieval you should set this parameter to `True`. This way the FileSimilarityRetriever can run on the CPU.
+
+# Detailed design - updated
+
+During the proposal discussion, it was decided to break up the initial FileSimilarityRetriever into two sub-components in order to simplify the design and testing and also to make it possible to use only one of the sub-components, if needed, thus covering more use cases.
+
+The new design would go as follows:
+- DocumentFetcher that gets all docs corresponding to a file based on a provided aggregation meta key and returns a list of docs to be used by one or two retrievers as input for batch retrieval.
+- DocumentAggregator placed after the retriever(s) that aggregates the document list(s) using RRF into a single ranked list of files, which are put together based on file_ids of retrieved docs.
+
+These sub-components will have the following params (for full explanation see Detailed design section above):
+- DocumentFetcher
+  - document_store
+  - file_aggregation_key
+  - max_num_docs: optional, in the original version it was called max_num_queries
+  Note: When researching how to change the params from the original implementation I found out there is no longer need for max_query_len that was required in the original FileSimilarityRetriever. It was implemented to not overshoot the max_clause_count limit in ElasticSearch settings but this got deprecated as of vs 8.0.0 according to their documentation - https://www.elastic.co/guide/en/elasticsearch/reference/current/search-settings.html
+
+- DocumentAggregator
+  - documents: can be a list or a list of list of docs
+  - file_aggregation_key: default would be "file_id", but can be changed to e.g. "name"
+  - output: default "top_document", but can also be "file_aggregation_key"
+  - keep_original_score
+  - top_k = 5
```

Comment:
@sjrl @MichelBartels I've updated the Detailed design section & the Basic example sections above - please take a look and let me know if sth can be improved 👍 

Also, for the implementation, I'm thinking what these sub-components should inherit from - one option could be DocumentFetcher based on BaseRetriever (bc in practice it will work similarly to FilterRetriever for example) and DocumentAggregator can be based on BaseRanker since it sort of ranks the retrieved documents by assessing full file similarity but not sure... WDYT?

Reply:
> one option could be DocumentFetcher based on BaseRetriever

Yeah, that makes sense to me. If it is going to basically be the FilterRetriever I do wonder if we should just use that node instead of creating a new one. What differences would there be between the FilterRetriever and the DocumentFetcher?

Reply:
> > one option could be DocumentFetcher based on BaseRetriever
> 
> Yeah, that makes sense to me. If it is going to basically be the FilterRetriever I do wonder if we should just use that node instead of creating a new one. What differences would there be between the FilterRetriever and the DocumentFetcher?

One small difference would be in how the metadata for filtering is provided - in the "filters" of the query (in FilterRetriever) or the meta key for aggregation is given as a param to DocumentFetcher and the value is received from the query itself. 
But this input part isn't as important as the output - DocumentFetcher needs to return a list of strings of fetched documents' contents (while FilterRetriever outputs haystack document objects), which can then be provided to the next retriever(s) as "queries" for batch retrieval at the next stage. Though this can be solved by adding an "output" param for FilterRetriever that makes it possible to get a list of strings instead of whole documents returned, what do you think @sjrl ?

Reply:
@elundaeva Sorry, I could only take a look today because I was on PTO last week. I would agree with Sebastian that it would be nicer if there wasn't a need for a separate `DocumentFetcher` node. However, I can see that it would have to be ugly either way because of the output type.

Have you also thought about batch file retrieval? It was originally supported by your draft PR, but I can't see how it's possible the way it's described in the current proposal. I also didn't consider batching when I suggested to split the components. I would guess if we were to enable batching somehow, it would be even harder to use the `FilterRetriever` because we would need to flatten the result at some point.